### DISCUSSION TITLE
in chat example, how to reuse `sender` as it is moved

### DISCUSSION THREAD

**Discussion by yatesco** [tokio-rs/axum#610] [Q&A] [answered]
Hi there, in line 94 https://github.com/tokio-rs/axum/blob/b1623ce7f2e0e58f9773a880d891590e564b2055/examples/chat/src/main.rs#L94 the `sender` is moved:
```rust
    // This task will receive broadcast messages and send text message to our client.
    let mut send_task = tokio::spawn(async move {
        while let Ok(msg) = rx.recv().await {
            // In any websocket error, break loop.
            if sender.send(Message::Text(msg)).await.is_err() {
                break;
            }
        }
    });
```

Because the sender can't be cloned, is it more idiomatic:
 - wrapping it an `rc` of some kind (feels weird)
 or
 - perform the `select!` of the various "business logic" channels (include the broadcast channel)?

I realise this is more a general design question - I'm just interested in the idioms here :-)

**Comment by davidpdrsn** [comment]
Putting the `WebSocket` in an arc wouldn't help because sending requires `&mut self` which you cannot get through an arc.

Is your use case that you need to send things on the socket from multiple places in your code concurrently?

**Comment by yatesco** [comment]
Exactly that.

**Comment by davidpdrsn** [comment, marked as answer]
I think I would make an mpsc channel and spawn a task to forward messages from the mpsc receiver to the socket sender. The mpsc sender can be cloned.

**Reply by yatesco** [reply]
Ah, got you. More and more I'm thinking of storing logic and state in the web socket loop itself is the wrong thing, and simply providing  a channel to the actual "session"/logic is more idiomatic.

**Reply by davidpdrsn** [reply]
Yep I agree. Channels are the best 😊
